src/expr/src/scalar.rs

```diff
@@ -1231,15 +1231,27 @@ impl MirScalarExpr {
 
                             // Defer errors until evaluation instead of eagerly returning them here
                             // to match the error behavior of the dynamic function (part of database-issues#4972).
-                            let regex = match func::build_regex(pattern, &flags) {
-                                Ok(regex) => Ok((regex, limit)),
-                                Err(err) => Err(err),
+                            *e = match func::build_regex(pattern, &flags) {
+                                Ok(regex) => {
+                                    let mut exprs = mem::take(exprs);
+                                    let replacement = exprs.swap_remove(2);
+                                    let source = exprs.swap_remove(0);
+                                    source.call_binary(
+                                        replacement,
+                                        BinaryFunc::RegexpReplace { regex, limit },
+                                    )
+                                }
+                                Err(err) => {
+                                    let mut exprs = mem::take(exprs);
+                                    let source = exprs.swap_remove(0);
+                                    let scalar_type = e.typ(column_types).scalar_type;
+                                    // We need to return `NULL` on `NULL` input, and error otherwise.
```

Comment:
Maybe extend the comment a bit. Something like:
"We'd like the same behaviors between the regex argument being dynamic or a literal. Therefore, we need to return `NULL` on `NULL` input, and error otherwise, because this is what the original, dynamic function does."